ako-gateway-api/nodes/dequeue_ingestion.go

```diff
@@ -1 +1,113 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
 package nodes
+
+import (
+	akogatewayapilib "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/ako-gateway-api/lib"
+	akogatewayapiobjects "github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/ako-gateway-api/objects"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/lib"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/nodes"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/objects"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+func DequeueIngestion(key string, fullsync bool) {
+	objType, namespace, name := lib.ExtractTypeNameNamespace(key)
+
+	schema, valid := ConfigDescriptor().GetByType(objType)
+	if !valid {
+		return
+	}
+	gatewayList, found := schema.GetGateways(name, namespace, key)
+	if !found {
+		//returning due to error, cannot delete or update
+		return
+	}
+	handleGateways(gatewayList, fullsync, key)
+}
+
+func handleGateways(gatewayList []string, fullsync bool, key string) {
+	sharedQueue := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
+	for _, gateway := range gatewayList {
+		_, namespace, name := lib.ExtractTypeNameNamespace(gateway)
+
+		modelName := lib.GetModelName(lib.GetTenant(), akogatewayapilib.GetGatewayParentName(namespace, name))
+		modelFound, _ := objects.SharedAviGraphLister().Get(modelName)
+
+		gatewayObj, err := akogatewayapilib.AKOControlConfig().GatewayApiInformers().GatewayInformer.Lister().Gateways(namespace).Get(name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				utils.AviLog.Infof("key: %s, got error while getting gateway class: %v", key, err)
+				continue
+			}
+			if modelFound {
+				objects.SharedAviGraphLister().Save(modelName, nil)
+				if !fullsync {
+					nodes.PublishKeyToRestLayer(modelName, key, sharedQueue)
+				}
+			}
+			akogatewayapiobjects.GatewayApiLister().DeleteGatewayToGatewayClass(namespace, name)
+			continue
+		}
+		gwClass := string(gatewayObj.Spec.GatewayClassName)
+		found, isAkoCtrl := akogatewayapiobjects.GatewayApiLister().IsGatewayClassControllerAKO(gwClass)
+		if !found {
+			//gateway class deleted
+			objects.SharedAviGraphLister().Save(modelName, nil)
+			if !fullsync {
+				nodes.PublishKeyToRestLayer(modelName, key, sharedQueue)
+			}
+			continue
+		}
+		if !isAkoCtrl {
+			//AKO is not the controller, do not build model
+			continue
+		}
+		aviModelGraph := NewAviObjectGraph()
+		aviModelGraph.BuildGatewayVs(gatewayObj, key)
+		if len(aviModelGraph.GetOrderedNodes()) > 0 {
+			ok := saveAviModel(modelName, aviModelGraph, key)
+			if ok && !fullsync {
+				nodes.PublishKeyToRestLayer(modelName, key, sharedQueue)
```

Comment:
please keep in my mind we have to add an info log here which marks the completion of graph layer processing. This will help us in debugging.

Reply:
added info log